### PR TITLE
IPv6 support

### PR DESCRIPTION
**Pull request by imoverclocked**
dbab previously failed to overwrite IPv6 addresses because dnsmasq only replaced IPv4 addresses with the specified 127.0.0.1:

before:

```
    % host 101order.com
    101order.com has address 127.0.0.1
    101order.com has IPv6 address 2606:4700:3035::ac43:82ac
    101order.com has IPv6 address 2606:4700:3032::6815:377
    101order.com mail is handled by 10 mail.rwgusa.net.
```

after:

```
    % host 101order.com
    101order.com has address 127.0.0.1
    101order.com has IPv6 address ::1
    101order.com mail is handled by 10 mail.rwgusa.net.
```

Tested on Debian host:
*  testing required installing libio-socket-inet6-perl
*  dash 0.5.11+git20210903+057cd650a4ed-9
*  dnsmasq 2.88-1
*  perl 5.36.0-6

### CODE REVIEW

**Comment by suntong** [comment]
Wow!! 

I noticed the problems given by IPv6 recently as well, but I had no knowledge/time of fixing it so I had to disable IPv6 on my system recently. Thanks a lot for such high quality fix Tim!

The changes look great but please give me some time, as I want to verify it myself before taking the PR, by installing it into a brand new Debian stable released just days ago.

**Comment by imoverclocked** [comment]
Thank you for creating this project!

**Comment by imoverclocked** [comment]
@suntong I noticed that you created+merged a parallel branch. Did you find something in testing that is problematic?

**Comment by suntong** [comment]
Oh, it's a way for me to test myself before merging it directly to master. I believe there would be no problem, just I'm still in the process of creating a clean hosting environment to test it. Stay tuned...

**Comment by suntong** [comment]
**IO-Socket-INET6 Deprecated**

See https://metacpan.org/dist/IO-Socket-INET6

> Deprecated.  
The maintainer of this distribution has indicated that it is deprecated and no longer suitable for use.

I did a quick search but wasn't able to find the _reason_ for the deprecation, and neither for the _alternative solutions_. 

@imoverclocked, might you have any idea?

**Comment by imoverclocked** [comment]
https://github.com/suntong/dbab/pull/11 should do the trick...

**Comment by suntong** [comment]
> % host 101order.com
>     101order.com has address 127.0.0.1
>     101order.com has IPv6 address ::1

Hmm... I have a noob question and I wasn't able to find it from the internet. So here is my situation:

```
$ host 101order.com
101order.com has address 10.0.5.5
101order.com has IPv6 address ::1
101order.com mail is handled by 10 mail.rwgusa.net.
```


- Suppose my GW is 10.0.5.1 and I have 3 local machines, from 10.0.5.2 to 10.0.5.4, and 10.0.5.5 is my DNS server.
- Would above DNS resolution work for _all_ of my local machines, originated from 10.0.5.2 to 10.0.5.4?
- Suppose my DNS server has the IP addresses of the following:

```
$ ip a s eth0
3: eth0@if5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1440 qdisc noqueue state UP group default 
    link/ether f6:ca:9e:f2:b8:62 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.0.5.2/30 brd 10.0.5.3 scope global eth0
       valid_lft forever preferred_lft forever
    inet 10.0.5.5/30 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::f4ca:9eff:fef2:b862/64 scope link 
       valid_lft forever preferred_lft forever
```

If I want to spell out the IPv6 of 10.0.5.5 in `/etc/dbab/dbab.addr` on the second line, what should I put there please?

Also, would you show the output from your side of the following command please?

`curl -vs --head https://s0.2mdn.net/`

Sorry again for such noob questions, and thanks for your help.

**Comment by suntong** [comment]
> If I want to spell out the IPv6 of 10.0.5.5 in `/etc/dbab/dbab.addr` on the second line, what should I put there please?

I did a `sudo ip a add ::5/64 dev eth0` after asking the above question and my new status is:

```
$ ip a s eth0
3: eth0@if5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1440 qdisc noqueue state UP group default 
    link/ether f6:ca:9e:f2:b8:62 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.0.5.2/30 brd 10.0.5.3 scope global eth0
       valid_lft forever preferred_lft forever
    inet 10.0.5.5/30 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 ::5/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::f4ca:9eff:fef2:b862/64 scope link 
       valid_lft forever preferred_lft forever
```

**Comment by imoverclocked** [comment]
IPv6 is subnetted similar to IPv4 but there are different ranges for different things.

::1 is defined as ::1/128 so there are no free bits like you might expect from 127.0.0.1/8

fe80::/10 is link-local only. Many systems require extra help to communicate anything above routing via link-local addresses and it is generally discouraged. DNS over link-local may not work (read: probably won’t for most scenarios you will try)

It looks like your router doesn’t have a proper IPv6 connection so the best you can do is something that is site-local or even something obscure like 6to4.

Basically: you need a “real” address to communicate with. I suppose the range you have added (::/64) might work if you configure everything on your LAN for it. Definitely non-standard though :)

**Comment by imoverclocked** [comment]
My local setup: I'm not worried about getting a pixel image back so I actually keep the resolution to loopback interfaces. My browser just sees "connection denied" since I don't run anything locally on port 80/443. Probably sending back nxdomain from dnsmasq would be a better use-case for me but the current setup works well enough.

**Comment by suntong** [comment]
Thanks for the detailed explanation.

> My browser just sees "connection denied" since I don't run anything locally on port 80/443.

OK. Hmm... I tried to replace the `::1` with my real IPv6 address, which is `ping`able from my local machines, but the pixel server seems to be kept crashing. I'll document how I test next, and hopefully you can find where the problem is...

**Comment by suntong** [comment]
> I'll document how I test next, and hopefully you can find where the problem is...

Done -- 
https://github.com/suntong/dbab-packer/wiki/dbab-within-Alpine-Linux-container#test-it-using-gitpod

in which you can see that the `server_ip_6` has been assigned twice.

- if not doing the last one and to have `server_ip_6` containing the real IPv6 address, then alpine work docker container will be kept crashing.
- if leaving the last one `server_ip_6='::1'` as-is, then DNS is OK, yet the pixel server seems not to be OK:

```sh
$ curl -s --head s0.2mdn.net
HTTP/1.0 200 OK
Server: dbab/v1.5
Connection: close
Cache-Control: public, max-age=31536000
Content-type: image/gif
Content-length: 43

$ curl -v6 -s --head s0.2mdn.net
*   Trying fe80::d872:d4ff:fe37:2c7f:80...
* Immediate connect fail for fe80::d872:d4ff:fe37:2c7f: Invalid argument
* Closing connection 0
```

**Comment by imoverclocked** [comment]
The problem is that you are using link local addresses:

`Immediate connect fail for fe80::d872:d4ff:fe37:2c7f: Invalid argument` 

*Invalid Argument* being the key.

Link local addresses are not meant for communication beyond low-level routing. Think of them more like a mac address.

You really need an actual IPv6 config for this to work properly. I don't recommend running production this way but you can choose a deprecated address scheme (site-local) just for your local testing:

DBAB-Server:
```
# ip -6 addr add fec0::5/64 dev eth0
# ip -6 addr show scope site
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fec0::5/64 scope site
       valid_lft forever preferred_lft forever
# (echo "10.0.5.5"; echo "fec0::5") > /etc/dbab/dbab.addr
# systemctl restart dbab
# dbab-get-list && systemctl restart dnsmasq
```

DBAB-Client: configure to use `fec0::6/64` and then you should be able to communicate with `fec0::5`. Validate with ping/nc/curl:

```
# ip -6 addr add fec0::6/64 dev end0
# ip -6 addr show scope site
2: end0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fec0::6/64 scope site
       valid_lft forever preferred_lft forever
# ping6 -c1 fec0::5
PING fec0::5(fec0::5) 56 data bytes
64 bytes from fec0::5: icmp_seq=1 ttl=64 time=0.965 ms

--- fec0::5 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.965/0.965/0.965/0.000 ms
#
# nc -6 -vz fec0::5 80
Connection to fec0::5 80 port [tcp/http] succeeded!
#
# curl -v --head http://s0.2mdn.net
*   Trying [fec0::5]:80...
* Connected to s0.2mdn.net (fec0::5) port 80 (#0)
> HEAD / HTTP/1.1
> Host: s0.2mdn.net
> User-Agent: curl/7.88.1
> Accept: */*
>
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
HTTP/1.0 200 OK
< Server: dbab/v1.5
Server: dbab/v1.5
< Connection: close
Connection: close
< Cache-Control: public, max-age=31536000
Cache-Control: public, max-age=31536000
< Content-type: image/gif
Content-type: image/gif
< Content-length: 43
Content-length: 43

<
* Excess found: excess = 43 url = / (zero-length body)
* Closing connection 0
```

**Comment by suntong** [comment]
Thanks a lot. All set now!!
